### ISSUE TITLE
Using Pointers to read in Model

### ISSUE THREAD

**Issue by carmesean** [swig/swig#1200]
I am attempting to read in a model to my module using a pointer. I have a module called Bar and a function named tell (in a separate class, Foo, which includes the Bar.h file): 

Foo.cpp

```
#include "time.h"
#include "stdio.h"
#include "stdlib.h"
#include "iostream"
#include "fstream"
#include "sstream"
#include "vector"
#include "set"
#include "map"
#include "string"
#include "string.h"
#include "Bar.h"
using namespace std;

void tell(vector <string> &a_in, Bar_Model *bar, vector <vector <double> > &b_out){
}
```

Bar.h (The Bar_Model class is within the Bar.h header file. I am only including the constructors) 
```
#pragma once
#include "iostream"
#include "fstream"
#include "sstream"
#include "string"
#include "vector"
#include "algorithm"
#include "cstdlib"
#include "cstdio"
#include "math"
#include "limits.h"  
using namespace std;

class Bar_Model;
class Bar_Model
{
public:
	 Bar_Model(int counter, string &str1, string &str2, int val1, int val2,
		int alpha=1,int beta =-1,double gamma=0,int eps=1);
	 Bar_Model(int alpha=1,int beta=-1,double gamma=0, int eps=1);
	~ Bar_Model();
};
```
Bar.i 

```
%module Bar
%{
#include "Bar.h"
%}

%include "std_string.i"
%include "std_vector.i"

// Instantiate templates used

%template(IntVector) std::vector<int>;
%template(IntVectorVector) std::vector< std::vector<int> >;
%template(DoubleVector) std::vector<double>;
%template(DoubleVectorVector) std::vector< std::vector<double> >;
%template(StringVector) std::vector<std::string>;
%template(StringVectorVector) std::vector< std::vector<std::string> >;
%template(ConstantCharacterVector) std::vector<const char*>;


/* Parse the original header file */
%include "Bar.h"
```

################

Foo.i 

```
%module Foo

%{
#include "Bar.h"
void tell(std::vector <std::string> &a_in, Bar_Model *bar, std::vector <std::vector <std::double> > &b_out);
%}

%include "std_string.i"
%include "std_vector.i"
using std::string;

// Instantiate templates used

%template(IntVector) std::vector<int>;
%template(IntVectorVector) std::vector< std::vector<int> >;
%template(DoubleVector) std::vector<double>;
%template(DoubleVectorVector) std::vector< std::vector<double> >;
%template(StringVector) std::vector<std::string>;
%template(StringVectorVector) std::vector< std::vector<std::string> >;
%template(ConstantCharacterVector) std::vector<const char*>;


#include "Bar.h"
void tell(std::vector <std::string> &a_in, Bar_Model *bar, std::vector <std::vector <std::double> > &b_out);

```
I then imported the modules and attempted to call the method tell:
```
>>> import Bar
>>> import Foo
>>> bar = "/Users/me/models/bar"
>>> Foo.tell(a_in, bar, b_out)
```

When I call this function in Python, I get the error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: in method 'tell', argument 2 of type 'Bar_Model *'
```

As I can not pass the model into the function directly, I am using the path to the directory. For example: `bar = "/Users/me/models/bar"`. I suspect I will need to include some type of typemap or apply directive for the Bar_Model type being as it is not a conventional variable type. How would I go about creating a type map for a class type, specifically in the case of my method "tell" and the Bar_Model object? 

####Using SWIG 3.0.12

(edited to include more information)

**Comment by ojwb** [comment]
You don't tell us anything at all about your `BarModel` type so it's hard for us to understand the situation fully.  I think you'll need to provide a short, self-contained example demonstrating the issue if you want anyone to be able to help.

**Comment by carmesean** [comment]
@ojwb I updated the issue with more information and clarified my problem!

**Comment by wsfulton** [comment]
> I think you'll need to provide a short, self-contained example demonstrating the issue if you want anyone to be able to help.
Your example isn't self contained. It's got syntax errors, it doesn't have an interface file. Try and recreate the output you have provided using just the information you have provided and then you'll understand that it is not self-contained.

**Comment by carmesean** [comment]
@wsfulton @ojwb sorry about that, I reedited the issue. I'm new to posting on forums, so I am still learning forum etiquette. I included the interface files for both Bar and Foo as well as corrected the errors. Please let me know if you run into another issue, I'll make sure to correct it.

**Comment by carmesean** [comment]
Also Foo.cpp is a class that contains a main method, which is why I do not have a header file for it, so I included the main class for greater context of the issue.

**Comment by ojwb** [comment]
As wsfulton pointed out to you already in #1198, you don't want `namespace std {` ... `}` around your `%template` uses and to also to explicitly qualify with `std::` in those `%template` uses.

**Comment by ojwb** [comment]
And there are still syntax errors, e.g. every include of `Bar.h` seems to use non-ASCII shaped double quotes, e.g.:

```
%include “Bar.h”
```

**Comment by carmesean** [comment]
@ojwb thanks for pointing out the quotes issue. I am not sure why it converted it to those types of quotes, but it is correct in my code. I removed the namespace, but the error remains.

**Comment by wsfulton** [comment]
Why have you got two different interface files, are you using multiple modules (that's an advanced topic where you need to provide a full build script)? Can you not distill the problem into one small example?

**Comment by carmesean** [comment]
@wsfulton Yes exactly! I am using multiple modules. The first module I am importing is `Bar`. The second module I am importing is `Foo`. `Foo` does a lot of the work required for the program using objects from `Bar`. Both are equally necessary for what I am trying to accomplish. I do not think I can condense the example any further without it seeming convoluted. 

Could you let me know what additional clarification is needed for the issue? I would be happy to provide any further details necessary. The root of the issue is my second argument for `tell` . The `BarModel *bar` parameter should be pointing to a model file in one of my folders. The model is of type `BarModel`, and `BarModel` is a nested class within the `Bar.h` file. I am calling that instance of the `BarModel` class in `Foo` via the function `tell`. 

I am importing them both. Importing multiple modules has worked successfully for me in another program in which I am also using instances of `Bar`. However, I think it is this particular parameter type that is causing me issues.

**Comment by ojwb** [comment]
I'm confused about `BarModel` vs `Bar_Model` - are those meant to be different?

**Comment by carmesean** [comment]
@ojwb No they are not, they are the same. I edited the issue and comment to reflect that

**Comment by carmesean** [comment]
@ojwb just checking back in. Do you have any suggestions on how to pass the class type in as a parameter? Would I have to use a type map?

**Comment by ojwb** [comment]
Sorry, I've given up looking at this issue - every time I do I seem to hit another problem which indicates the example you're showing us isn't the one you're testing locally, and that's just abusing my good will and wasting my time.

It might be `%implicitconv` would help you.

**Comment by carmesean** [comment]
@ojwb I apologize for the issue I had with documenting the issue. As I mentioned, I am new to SWIG and new to forums. I am college student just trying to use a software tool I thought was really cool and innovative.  I did not realize the quotation notation was a syntax error as it works fine in my own code. Regarding the example I am showing, it is in fact the one I am testing locally. I refined it to make it a more general example of the issue I am having. Everything is the same. Regarding the difference between BarModel and Bar_Model, that was a mistake I made as I was refining the example.
